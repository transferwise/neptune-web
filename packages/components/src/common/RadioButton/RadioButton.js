import React from 'react';
import Types from 'prop-types';
import classNames from 'classnames';

import './RadioButton.css';

const RadioButton = ({ id, value, name, checked, onChange, disabled, readOnly }) => (
  <>
    <input
      type="radio"
      className="sr-only"
      id={id}
      value={value}
      name={name}
      checked={checked}
      onChange={() => (!checked ? onChange(value) : null)}
      disabled={disabled || readOnly}
    />
    <span className={classNames('tw-radio-button', { checked, disabled: disabled || readOnly })}>
      <span className="tw-radio-check" />
    </span>
  </>
);

RadioButton.propTypes = {
  id: Types.string,
  name: Types.string.isRequired,
  checked: Types.bool,
  onChange: Types.func,
  disabled: Types.bool,
  value: Types.oneOfType([Types.number, Types.string]),
  readOnly: Types.bool,
};

RadioButton.defaultProps = {
  checked: false,
  onChange: () => {},
  disabled: false,
  id: null,
  value: '',
  readOnly: false,
};

export default RadioButton;
