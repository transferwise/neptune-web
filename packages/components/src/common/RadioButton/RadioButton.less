input[type='radio'] {
  &:focus,
  &:active {
    & + .tw-radio-button {
      border-color: var(--color-border-focus);
      outline: none;
    }
  }
}

.tw-radio-button.disabled {
  background-color: var(--color-background-neutral);
}

.radio:hover .tw-radio-button.disabled {
  border-color: var(--color-border-neutral);
}
