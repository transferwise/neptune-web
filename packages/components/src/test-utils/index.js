import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
