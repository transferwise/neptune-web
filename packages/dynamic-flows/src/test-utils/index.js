import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
