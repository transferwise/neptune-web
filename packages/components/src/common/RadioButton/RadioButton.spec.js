import React from 'react';
import { screen, render, userEvent } from '../../test-utils';

import RadioButton from '.';

describe('Radio button', () => {
  it('renders correctly', () => {
    const { container } = render(<RadioButton id="an-id" name="a-name" value="a-value" />);

    expect(container).toMatchSnapshot();
  });

  it('renders checked state correctly', () => {
    const { container } = render(<RadioButton name="radio-name" checked />);

    expect(container).toMatchSnapshot();
  });

  it('renders disabled state correctly', () => {
    const { container } = render(<RadioButton name="radio-name" disabled />);

    expect(container).toMatchSnapshot();
  });

  it('responds to user input', () => {
    const onChange = jest.fn();
    render(
      <label htmlFor="an-id">
        <RadioButton onChange={onChange} id="an-id" name="a-name" value="a-value" />A radio
      </label>,
    );

    userEvent.click(screen.getByLabelText('A radio'));

    expect(onChange).toHaveBeenCalledWith('a-value');
  });
});
