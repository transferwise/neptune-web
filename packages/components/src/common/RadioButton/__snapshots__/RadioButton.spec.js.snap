// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio button renders checked state correctly 1`] = `
<div>
  <input
    checked=""
    class="sr-only"
    name="radio-name"
    type="radio"
    value=""
  />
  <span
    class="tw-radio-button checked"
  >
    <span
      class="tw-radio-check"
    />
  </span>
</div>
`;

exports[`Radio button renders correctly 1`] = `
<div>
  <input
    class="sr-only"
    id="an-id"
    name="a-name"
    type="radio"
    value="a-value"
  />
  <span
    class="tw-radio-button"
  >
    <span
      class="tw-radio-check"
    />
  </span>
</div>
`;

exports[`Radio button renders disabled state correctly 1`] = `
<div>
  <input
    class="sr-only"
    disabled=""
    name="radio-name"
    type="radio"
    value=""
  />
  <span
    class="tw-radio-button disabled"
  >
    <span
      class="tw-radio-check"
    />
  </span>
</div>
`;
